transport/ProbeServer.cpp

```diff
@@ -0,0 +1,224 @@
+    memory::UniquePacket packet)
+{
+}
+
+void ProbeServer::onIceReceived(Endpoint& endpoint,
+    const SocketAddress& source,
+    const SocketAddress& target,
+    memory::UniquePacket packet)
+{
+    replyStunOk(endpoint, source, std::move(packet));
+}
+
+void ProbeServer::onRegistered(Endpoint& endpoint)
+{
+    const std::lock_guard<std::mutex> locker(_endpointsLock);
+
+    if (endpoint.getTransportType() != ice::TransportType::UDP)
+    {
+        return;
+    }
+
+    auto port = endpoint.getLocalPort().getPort();
+    auto it = _udpPorts.find(port);
+    if (it != _udpPorts.end())
+    {
+        return;
+    }
+
+    if (!_iceConfig.publicIpv4.empty() && endpoint.getLocalPort().getFamily() == AF_INET)
+    {
+        auto address = _iceConfig.publicIpv4;
+        address.setPort(port);
+        _probeCandidates.emplace_back(ProbeCandidate{ice::TransportType::UDP, address});
+
+        _udpPorts.insert(port);
+
+        logger::debug("Registered UDP endpoint %s:%u", _name, address.ipToString().c_str(), port);
+    }
+
+    if (!_iceConfig.publicIpv6.empty() && endpoint.getLocalPort().getFamily() == AF_INET6)
+    {
+        auto address = _iceConfig.publicIpv6;
+        address.setPort(port);
+        _probeCandidates.emplace_back(ProbeCandidate{ice::TransportType::UDP, address});
+
+        _udpPorts.insert(port);
+
+        logger::debug("Registered UDP endpoint %s:%u", _name, address.ipToString().c_str(), port);
+    }
+}
+
+void ProbeServer::onUnregistered(Endpoint& endpoint)
+{
+    const std::lock_guard<std::mutex> locker(_endpointsLock);
+
+    if (endpoint.getTransportType() == ice::TransportType::UDP)
+    {
+        auto port = endpoint.getLocalPort().getPort();
+        _udpPorts.erase(port);
+
+        std::remove_if(_probeCandidates.begin(), _probeCandidates.end(), [port](const transport::ProbeCandidate& c) {
+            return c.transport == ice::TransportType::UDP && c.address.getPort() == port;
+        });
+    }
+}
+
+// ServerEndpoint::IEvents
+void ProbeServer::onServerPortRegistered(ServerEndpoint& endpoint)
+{
+    const std::lock_guard<std::mutex> locker(_endpointsLock);
+
+    auto port = endpoint.getLocalPort().getPort();
+    auto it = _tcpPorts.find(port);
+    if (it != _tcpPorts.end())
+    {
+        return;
+    }
+
+    if (!_iceConfig.publicIpv4.empty() && endpoint.getLocalPort().getFamily() == AF_INET)
+    {
+        auto address = _iceConfig.publicIpv4;
+        address.setPort(port);
+        _probeCandidates.emplace_back(ProbeCandidate{ice::TransportType::TCP, address});
+
+        _tcpPorts.insert(port);
+
+        logger::debug("Registered TCP server endpoint %s:%u", _name, address.ipToString().c_str(), port);
+    }
+
+    if (!_iceConfig.publicIpv6.empty() && endpoint.getLocalPort().getFamily() == AF_INET6)
+    {
+        auto address = _iceConfig.publicIpv6;
+        address.setPort(port);
+        _probeCandidates.emplace_back(ProbeCandidate{ice::TransportType::TCP, address});
```

Comment:
I think emplace_back can even take the constructor arguments directly for the ProbeCandidate.

Reply:
Fixed.